File: 201902769/17387.py
import sys
input = sys.stdin.readline

def ccw(p1, p2, p3): #CCW
    x1, y1 = p1
    x2, y2 = p2
    x3, y3 = p3
    return (x1 * y2 + x2 * y3 + x3 * y1) - (x2 * y1 + x3 * y2 + x1 * y3)

x1, y1, x2, y2 = map(int, input().split()) #선분 A
x3, y3, x4, y4 = map(int, input().split()) #선분 B
p1, p2, p3, p4 = [x1, y1], [x2, y2], [x3, y3], [x4, y4]
cross1 = ccw(p1, p2, p3) * ccw(p1, p2, p4) #선분 A에서 B에 대한 CCW 값
cross2 = ccw(p3, p4, p1) * ccw(p3, p4, p2) #선분 B에서 A에 대한 CCW 값
ans = 0 #기본값 (교차하지 않음)
if cross1 == cross2 == 0: #모두 0이라면 (한 직선 위에 있으면)
    if min(x1, x2) <= max(x3, x4) and min(x3, x4) <= max(x1, x2) and min(y1, y2) <= max(y3, y4) and min(y3, y4) <= max(y1, y2): #두 직선이 겹치면 (점끼리 만나는 것도 포함)
        ans = 1 #교차
elif cross1 <= 0 and cross2 <= 0: #모두 음수라면
    ans = 1 #교차
print(ans)

File: 201902769/17387.rs
use std::io;

struct Line {
    a: (i128, i128),
    b: (i128, i128),
}

impl Line {
    fn new(a: (i128, i128), b: (i128, i128)) -> Self {
        Self { a, b }
    }

    fn ccw(&self, c: (i128, i128)) -> i128 {
        let (ax, ay) = self.a;
        let (bx, by) = self.b;
        let (cx, cy) = c;
        (bx - ax) * (cy - ay) - (by - ay) * (cx - ax)
    }

    fn within(&self, c: (i128, i128)) -> bool {
        let (ax, ay) = self.a;
        let (bx, by) = self.b;
        let (cx, cy) = c;
        let min_x = ax.min(bx);
        let max_x = ax.max(bx);
        let min_y = ay.min(by);
        let max_y = ay.max(by);
        cx >= min_x && cx <= max_x && cy >= min_y && cy <= max_y
    }
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.trim().split_whitespace();
    let line1 = Line::new(
        (iter.next().unwrap().parse().unwrap(), iter.next().unwrap().parse().unwrap()),
        (iter.next().unwrap().parse().unwrap(), iter.next().unwrap().parse().unwrap()),
    );

    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.trim().split_whitespace();
    let line2 = Line::new(
        (iter.next().unwrap().parse().unwrap(), iter.next().unwrap().parse().unwrap()),
        (iter.next().unwrap().parse().unwrap(), iter.next().unwrap().parse().unwrap()),
    );

    let ccw1 = line1.ccw(line2.a) * line1.ccw(line2.b);
    let ccw2 = line2.ccw(line1.a) * line2.ccw(line1.b);

    // ccw1, ccw2 둘 다 0인 경우는 두 선분이 일직선 상에 있는 경우
    if ccw1 == 0 && ccw2 == 0 {
        // 두 선분이 일직선 상에 있는 경우, 두 선분이 겹치는지 확인
        // 4개 다 비교하는 이유는 포함 관계인 경우도 있기 때문
        if line1.within(line2.a) || line1.within(line2.b) || line2.within(line1.a) || line2.within(line1.b){
            println!("{}", 1);
        } else {
            println!("{}", 0);
        }
    }
    else if ccw1 <= 0 && ccw2 <= 0 {
        println!("{}", 1);
    } else {
        println!("{}", 0);
    }
}
